lib/pbench/server/sync.py

```diff
@@ -27,116 +19,156 @@ def __str__(self) -> str:
-        atomic. This seems to be difficult to achieve in a manner that works
-        both with PostgreSQL for production and sqlite3 for unit tests. When
-        we've removed the use of sqlite3 we can make this a nested transaction.
+        """Advertise the operations for which the dataset is now ready.
 
         Args:
             dataset: The dataset
-            did: The operation (if any) just completed, which will be removed
-                from the list of eligible operations.
+            did: The new OperationState of the component Operation
             enabled: A list (if any) of operations for which the dataset is now
                 eligible.
-            status: A status message (if not specified, and enabling new
-                operation(s), the default is "ok")
+            message: An optional status message
         """
-        ops: JSONVALUE = Metadata.getvalue(dataset, Metadata.OPERATION)
-        message = status
-
-        if not ops:
-            operations = set()
-        else:
-            operations = set(ops)
-            if did:
-                operations.discard(did.name)
-
-        # If we "did" something, the default message is "ok"
-        if did and not message:
-            message = "ok"
 
+        if enabled is None:
+            enabled = []
+
+        self.logger.debug(
+            "Dataset {} did {}, enabling {} with message {!r}",
+            dataset.name,
+            did.name if did else "nothing",
+            [e.name for e in enabled] if enabled else "none",
+            message,
+        )
+
+        # Don't reference the Dataset (which is outside our session scope)
+        # inside the context manager.
+        ds_name = dataset.name
+        ds_id = dataset.id
+        retries = self.RETRIES
+
+        # To avoid SELECT statements after updates, and the whole complication
+        # of SQLAlchemy AUTOFLUSH semantics, we're going to gather a set of all
+        # the operation rows we might need to update at the beginning.
+        match_set: set[OperationName] = set()
+        if did or message:
+            match_set.add(self.component)
         if enabled:
-            operations.update(o.name for o in enabled)
-
-        try:
-            self.logger.info(
-                "Did {}, enabling {} with message {!r}",
-                did.name if did else "nothing",
-                [e.name for e in enabled] if enabled else "none",
-                message,
-            )
-            Metadata.setvalue(dataset, Metadata.OPERATION, sorted(operations))
-            if message:
-                Metadata.setvalue(dataset, "server.status." + self.component, message)
-        except Exception as e:
-            self.logger.warning("{} error updating ops: {}", dataset.name, str(e))
-            raise
+            match_set.update(enabled)
+
+        filters = [Operation.name == n for n in match_set]
+
+        while retries > 0:
```

Comment:
Maybe I'll try this later. I don't want to mess with anything deep because I'm having a lot of trouble running outside the unit tests. I can bring up a pod and container on my VM, but the server didn't even start until I hacked `pbench-server.cfg` to correct the realm name (set up by `load_keycloak.sh` as `pbench-server` but defined in my container `pbench-server.cfg` somehow as just `pbench`; and although I'm sure I've seen the code at some point where that's generated, I can't find it now). Even once I fix that, the functional tests are having trouble logging in. Today's a mess, and I just kinda want to get the basic transaction bug fixed and move on.

Reply:
A recent update to the `main` branch changed the canned `pbench-server.cfg` to remove the realm configuration altogether, and it changes the default realm in `load_keycloak.sh` to `pbench-server`.

Reply:
Yeah, figured it was something like that. It's a pain to start from `main` when I'm doing DB changes, and I had to rejigger my familiar pod setup for the recent changes. I started to set up a minimal development environment so I could pull a branch and build containers on my VM, but I keep running into implicit dependencies and I got too impatient to get through them all today. Maybe later.